src/libp2p.ts

```diff
@@ -139,124 +115,97 @@ export class Libp2pNode extends EventEmitter<Libp2pEvents> implements Libp2p {
-        this.onDiscoveryPeer(evt)
-      })
-    }
-
-    this.peerRouting = this.components.peerRouting = this.configureComponent(new DefaultPeerRouting(this.components, {
+    // Peer routers
+    const peerRouters: PeerRouting[] = (init.peerRouters ?? []).map((fn, index) => this.configureComponent(`peer-router-${index}`, fn(this.components)))
+    this.peerRouting = this.components.peerRouting = this.configureComponent('peerRouting', new DefaultPeerRouting(this.components, {
       ...init.peerRouting,
       routers: peerRouters
     }))
 
-    const contentRouters: ContentRouting[] = (init.contentRouters ?? []).map(fn => this.configureComponent(fn(this.components)))
-
-    if (init.dht != null) {
-      // add dht to routers
-      contentRouters.push(this.configureComponent(new DHTContentRouting(this.dht)))
-    }
-
-    this.contentRouting = this.components.contentRouting = this.configureComponent(new CompoundContentRouting(this.components, {
+    // Content routers
+    const contentRouters: ContentRouting[] = (init.contentRouters ?? []).map((fn, index) => this.configureComponent(`content-router-${index}`, fn(this.components)))
+    this.contentRouting = this.components.contentRouting = this.configureComponent('contentRouting', new CompoundContentRouting(this.components, {
       routers: contentRouters
     }))
 
-    this.fetchService = this.configureComponent(new FetchService(this.components, {
-      ...init.fetch
-    }))
-
-    this.pingService = this.configureComponent(new PingService(this.components, {
-      ...init.ping
-    }))
-
-    this.autonatService = this.configureComponent(new AutonatService(this.components, {
-      ...init.autonat
-    }))
-
-    if (init.relay != null) {
-      this.circuitService = this.configureComponent(init.relay(this.components))
-    }
-
     // Discovery modules
-    for (const fn of init.peerDiscovery ?? []) {
-      const service = this.configureComponent(fn(this.components))
+    ;(init.peerDiscovery ?? []).forEach((fn, index) => {
+      const service = this.configureComponent(`peer-discovery-${index}`, fn(this.components))
 
       service.addEventListener('peer', (evt) => {
-        this.onDiscoveryPeer(evt)
+        this.#onDiscoveryPeer(evt)
       })
-    }
+    })
 
     // Transport modules
-    init.transports.forEach((fn) => {
-      this.components.transportManager.add(this.configureComponent(fn(this.components)))
+    init.transports.forEach((fn, index) => {
+      this.components.transportManager.add(this.configureComponent(`transport-${index}`, fn(this.components)))
     })
-  }
 
-  private configureComponent <T> (component: T): T {
-    if (isStartable(component)) {
-      this.services.push(component)
+    // User defined modules
+    if (init.services != null) {
+      for (const name of Object.keys(init.services)) {
+        const createService = init.services[name]
+        const service: any = createService(this.components)
+
+        if (service == null) {
```

Comment:
maybe deserves a warn/trace log here.
maybe also in `configureComponent`

Reply:
Done!